plugins/radar-android-google-places/src/main/java/org.radarbase.passive.google.places/GooglePlacesManager.kt

```diff
@@ -0,0 +1,454 @@
+            logger.error("Exception while initializing places API", ex)
+            service.internalError.set(true)
+            disconnect()
+        }
+    }
+
+    @Synchronized
+    @SuppressLint("MissingPermission")
+    private fun processPlacesData() {
+        placeHandler.execute {
+            if (isPermissionGranted) {
+                val client = placesClient ?: return@execute
+                currentPlaceRequest = FindCurrentPlaceRequest.newInstance(currentPlaceFields)
+                   // resetting the backoff time in SharedPreferences once the plugin works successfully
+                if (service.numOfAttempts > 0) {
+                        service.preferences.edit()
+                            .putInt(NUMBER_OF_ATTEMPTS_KEY, 0).apply()
+                    }
+                    client.findCurrentPlace(currentPlaceRequest)
+                        .addOnSuccessListener { response: FindCurrentPlaceResponse ->
+                            val limitLikelihood = limitByPlacesLikelihood
+                            val countLikelihood = limitByPlacesCount
+                            val placeLikelihoods = response.placeLikelihoods.toMutableList()
+                            if (limitLikelihood >= 0.0)  {
+                                placeLikelihoods.removeAll { it.likelihood < limitLikelihood }
+                            }
+                            placeLikelihoods.sortByDescending { it.likelihood }
+                            doSend(
+                                if (countLikelihood >= 0) {
+                                    placeLikelihoods.take(countLikelihood)
+                                } else {
+                                    placeLikelihoods
+                                }
+                            )
+                        }
+                        .addOnFailureListener { ex ->
+                            when (ex.javaClass) {
+                                ApiException::class.java -> {
+                                    val exception = ex as ApiException
+                                    handleApiException(exception, exception.statusCode)
+                                }
+                                else -> logger.error("Exception while fetching current place", ex)
+                            }
+                        }
+                } else {
+                    logger.warn("Location permissions not granted yet")
+                }
+            }
+        }
+
+    private fun doSend(places: List<PlaceLikelihood>) {
+        places.forEach { likelihood ->
+            val types = likelihood.place.types?.map { it.toPlacesType() }
+            val type1 = types?.getOrNull(0)
+            val type2 = types?.getOrNull(1)
+            val type3 = types?.getOrNull(2)
+            val type4 = types?.getOrNull(3)
+            val placeLikelihood: Double = likelihood.likelihood
+            val placeId: String? = if (shouldFetchPlaceId) likelihood.place.id else null
```

Comment:
Instead of putting all of this in separate variables, you could use a builder instead
```kotlin
val placesInfoBuilder = GooglePlacesInfo.Builder().apply {
    time = currentTime
    timeReceived = currentTime
    type1 = types?.getOrNull(0)
    ...
}
```
and then in all the branches you just have to add any additional components if applicable and then
```kotlin
send(..., placesInfoBuilder.build())
```